tsdb/engine/tsm1/iterator.gen.go

```diff
@@ -0,0 +1,1081 @@
+}
+
+func newBooleanIterator(name string, tags influxql.Tags, opt influxql.IteratorOptions, cur booleanCursor, aux []cursorAt, conds []*bufCursor, condNames []string) *booleanIterator {
+	itr := &booleanIterator{
+		cur: cur,
+		aux: aux,
+		opt: opt,
+		point: influxql.BooleanPoint{
+			Name: name,
+			Tags: tags,
+		},
+	}
+
+	if len(aux) > 0 {
+		itr.point.Aux = make([]interface{}, len(aux))
+	}
+
+	if opt.Condition != nil {
+		itr.m = make(map[string]interface{}, len(aux)+len(conds))
+	}
+	itr.conds.names = condNames
+	itr.conds.curs = conds
+
+	return itr
+}
+
+// Next returns the next point from the iterator.
+func (itr *booleanIterator) Next() *influxql.BooleanPoint {
+	for {
+		seek := tsdb.EOF
+
+		if itr.cur != nil {
+			// Read from the main cursor if we have one.
+			itr.point.Time, itr.point.Value = itr.cur.nextBoolean()
+			seek = itr.point.Time
+		} else {
+			// Otherwise find lowest aux timestamp.
+			for i := range itr.aux {
+				if k, _ := itr.aux[i].peek(); seek == tsdb.EOF || k < seek {
+					seek = k
+				}
+			}
+			itr.point.Time = seek
+		}
+
+		// Exit if we have no more points or we are outside our time range.
+		if itr.point.Time == tsdb.EOF {
+			return nil
+		} else if itr.opt.Ascending && itr.point.Time > itr.opt.EndTime {
+			return nil
+		} else if !itr.opt.Ascending && itr.point.Time < itr.opt.StartTime {
+			return nil
+		}
+
+		// Read from each auxiliary cursor.
+		for i := range itr.opt.Aux {
+			itr.point.Aux[i] = itr.aux[i].nextAt(seek)
+		}
+
+		// Read from condition field cursors.
+		for i := range itr.conds.curs {
+			itr.m[itr.conds.names[i]] = itr.conds.curs[i].nextAt(seek)
+		}
+
+		// Evaluate condition, if one exists. Retry if it fails.
+		if itr.opt.Condition != nil && !influxql.EvalBool(itr.opt.Condition, itr.m) {
+			continue
+		}
+
+		return &itr.point
+	}
+}
+
+// Close closes the iterator.
+func (itr *booleanIterator) Close() error { return nil }
+
+// booleanCursor represents an object for iterating over a single boolean field.
+type booleanCursor interface {
+	cursor
+	nextBoolean() (t int64, v bool)
+}
+
+func newBooleanCursor(seek int64, ascending bool, cacheValues Values, tsmKeyCursor *KeyCursor) booleanCursor {
+	if ascending {
+		return newBooleanAscendingCursor(seek, cacheValues, tsmKeyCursor)
+	}
+	return newBooleanDescendingCursor(seek, cacheValues, tsmKeyCursor)
+}
+
+type booleanAscendingCursor struct {
+	cache struct {
+		values Values
+		pos    int
+	}
+
+	tsm struct {
+		buf       []*BooleanValue
+		values    []*BooleanValue
+		pos       int
+		keyCursor *KeyCursor
+	}
+}
+
+func newBooleanAscendingCursor(seek int64, cacheValues Values, tsmKeyCursor *KeyCursor) *booleanAscendingCursor {
+	c := &booleanAscendingCursor{}
+
+	c.cache.values = cacheValues
+	c.cache.pos = sort.Search(len(c.cache.values), func(i int) bool {
+		return c.cache.values[i].Time().UnixNano() >= seek
+	})
+
+	c.tsm.keyCursor = tsmKeyCursor
+	c.tsm.buf = make([]*BooleanValue, 1000)
```

Comment:
I ended up removing some allocations like this from KeyCursor because they blew up the heap with larger queries.